compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs

```diff
@@ -192,18 +192,23 @@ envInsertAlias v x env = env{envAliases = MS.insert v x (envAliases env)}
 envLookupAlias :: Var -> Env -> Maybe LF.Expr
 envLookupAlias x = MS.lookup x . envAliases
 
--- | Bind a type var without shadowing its LF name.
-envBindTypeVar :: Var -> Env -> (TypeVarName, Env)
-envBindTypeVar x env = try 1 (TypeVarName occName)
+-- | Get a fresh type var whose LF name has not yet been taken.
+envFreshTypeVar :: T.Text -> Env -> (TypeVarName, Env)
+envFreshTypeVar nameHint env = try 1 (TypeVarName nameHint)
     where
-        occName = getOccText x
-        nameFor i = TypeVarName (occName <> T.pack (show i))
+        nameFor i = TypeVarName (nameHint <> T.pack (show i))
 
         try :: Int -> TypeVarName -> (TypeVarName, Env)
         try !i name =
             if envHasTypeVarName name env
                 then try (i+1) (nameFor i)
-                else (name, envInsertTypeVar x name env)
+                else (name, env{envTypeVarNames = S.insert name (envTypeVarNames env)})
+
+-- | Bind a type var without shadowing its LF name.
+envBindTypeVar :: Var -> Env -> (TypeVarName, Env)
+envBindTypeVar x env0 =
+    let (n, env) = envFreshTypeVar (getOccText x) env in
```

Comment:
Hmm. I wonder what this does.

```suggestion
    let (n, env) = envFreshTypeVar (getOccText x) env0 in
```

Reply:
Excellent catch! Looks like the giant foot gun that `let`s are recursive by default just went off. I wonder what the designers of Haskell thought when they made that design decision? That's maybe my biggest pet peeve about Haskell.